File: src/lib/rbac.ts
export type Role = 'student' | 'staff' | 'admin' | 'super_admin' | 'principal' | 'director' | 'hod' | 'head_warden' | 'warden' | 'incharge' | 'chef' | 'head_chef' | 'gate_security' | 'security_head' | 'hr' | 'pd' | 'pt'

// Role constants - must match backend exactly
export const ROLE_STUDENT = 'student'
export const ROLE_STAFF = 'staff'
export const ROLE_ADMIN = 'admin'
export const ROLE_SUPER_ADMIN = 'super_admin'
export const ROLE_PRINCIPAL = 'principal'
export const ROLE_DIRECTOR = 'director'
export const ROLE_HOD = 'hod'
export const ROLE_HEAD_WARDEN = 'head_warden'
export const ROLE_WARDEN = 'warden'
export const ROLE_INCHARGE = 'incharge'
export const ROLE_CHEF = 'chef'
export const ROLE_HEAD_CHEF = 'head_chef'
export const ROLE_GATE_SECURITY = 'gate_security'
export const ROLE_SECURITY_HEAD = 'security_head'
export const ROLE_PD = 'pd'
export const ROLE_PT = 'pt'
export const ROLE_HR = 'hr'

// Role groups
export const ADMIN_ROLES = [ROLE_SUPER_ADMIN, ROLE_ADMIN]
export const AUTHORITY_ROLES = [ROLE_SUPER_ADMIN, ROLE_ADMIN, ROLE_HEAD_WARDEN, ROLE_WARDEN]
export const STAFF_ROLES = [ROLE_SUPER_ADMIN, ROLE_ADMIN, ROLE_PRINCIPAL, ROLE_DIRECTOR, ROLE_HOD, ROLE_HEAD_WARDEN, ROLE_WARDEN, ROLE_INCHARGE, ROLE_STAFF, ROLE_CHEF, ROLE_HEAD_CHEF, ROLE_PD, ROLE_PT, ROLE_HR]
export const SECURITY_ROLES = [ROLE_SUPER_ADMIN, ROLE_ADMIN, ROLE_SECURITY_HEAD, ROLE_GATE_SECURITY]
export const WARDEN_ROLES = [ROLE_SUPER_ADMIN, ROLE_ADMIN, ROLE_HEAD_WARDEN, ROLE_WARDEN]
export const CHEF_ROLES = [ROLE_CHEF, ROLE_HEAD_CHEF]
export const GATE_ROLES = [ROLE_GATE_SECURITY, ROLE_SECURITY_HEAD]
export const MANAGEMENT_ROLES = [...AUTHORITY_ROLES, ROLE_STAFF, ROLE_PRINCIPAL, ROLE_DIRECTOR, ROLE_HOD, ROLE_INCHARGE]
export const TOP_LEVEL_ROLES = [ROLE_SUPER_ADMIN, ROLE_ADMIN, ROLE_HEAD_WARDEN]

// Role hierarchy weights
export const ROLE_HIERARCHY: Record<Role, number> = {
  student: 0,
  staff: 40,
  warden: 50,
  gate_security: 25,
  chef: 20,
  head_warden: 60,
  security_head: 35,
  head_chef: 30,
  incharge: 55,
  hod: 70,
  principal: 80,
  director: 75,
  hr: 45,
  pd: 20,
  pt: 15,
  admin: 90,
  super_admin: 100,
}

// Helper functions
export const isAdmin = (role?: string | null) => role ? ADMIN_ROLES.includes(role as Role) : false
export const isStaff = (role?: string | null) => role ? STAFF_ROLES.includes(role as Role) : false
export const isStudent = (role?: string | null) => role === ROLE_STUDENT
export const isWarden = (role?: string | null) => role ? WARDEN_ROLES.includes(role as Role) : false
export const isSecurity = (role?: string | null) => role ? SECURITY_ROLES.includes(role as Role) : false
export const isManagement = (role?: string | null) => role ? MANAGEMENT_ROLES.includes(role as Role) : false
export const isTopLevelManagement = (role?: string | null) => role ? TOP_LEVEL_ROLES.includes(role as Role) : false

/**
 * Check if the first role is strictly higher than the second role in the hierarchy.
 */
export const isHigherRole = (role1: string | null | undefined, role2: string | null | undefined): boolean => {
  if (!role1 || !role2) return false
  const r1 = role1.toLowerCase() as Role
  const r2 = role2.toLowerCase() as Role
  return (ROLE_HIERARCHY[r1] || 0) > (ROLE_HIERARCHY[r2] || 0)
}

/**
 * Check if a user can edit another user based on hierarchy.
 * Admin/Super Admin can edit anyone.
 * Others can only edit if they are higher in hierarchy and in the same branch (handled via logic in UI or backend).
 */
export const canEditUser = (currentUserRole: string | null | undefined, targetUserRole: string | null | undefined): boolean => {
   if (!currentUserRole || !targetUserRole) return false
   if (isAdmin(currentUserRole)) return true
   return isHigherRole(currentUserRole, targetUserRole)
}

export const ROLE_HOME: Record<Role, string> = {
  student: '/dashboard',
  staff: '/dashboard',
  admin: '/dashboard',
  super_admin: '/colleges',
  principal: '/reports',
  director: '/reports',
  hod: '/events',
  head_warden: '/dashboard',
  warden: '/dashboard',
  incharge: '/dashboard',
  chef: '/meals',
  head_chef: '/meals',
  gate_security: '/gate-passes',
  security_head: '/gate-passes',
  hr: '/dashboard',
  pd: '/sports-dashboard',
  pt: '/sports-dashboard',
}

export const COMMON_PATHS = [
    '/dashboard', '/profile', '/notifications', '/messages', '/notices', '/digital-id', '/unauthorized'
]


const ROLE_ALLOWED_PATHS: Record<Role, string[]> = {
  student: [
    ...COMMON_PATHS,
    '/rooms',
    '/attendance',
    '/meals',
    '/leaves',
    '/visitors',
    '/complaints',
    '/gate-passes',
    '/fines',
    '/events',
    '/sports-booking',
    '/resume',
    '/hall-booking',
    '/placements',
    '/room-requests',
  ],
  staff: [
    ...COMMON_PATHS,
    '/sports-booking',
  ],
  warden: [
    ...COMMON_PATHS,
    '/rooms',
    '/gate-passes',
    '/attendance',
    '/meals',
    '/gate-scans',
    '/tenants',
    '/metrics',
    '/reports',
    '/room-mapping',
    '/complaints',
    '/leaves',
    '/visitors',
    '/fines',
    '/sports-booking',
  ],
  head_warden: [
    ...COMMON_PATHS,
    '/rooms', '/gate-passes', '/attendance', '/meals',
    '/tenants', '/metrics', '/reports', '/room-mapping', '/complaints',
    '/leaves', '/visitors', '/room-requests',
    '/events', '/fines', '/analytics', '/hall-booking', '/sports-booking', '/audit-logs',
  ],
  admin: [
    ...COMMON_PATHS,
    '/rooms', '/gate-passes', '/attendance', '/meals',
    '/events', '/complaints', '/visitors',
    '/gate-scans', '/colleges', '/tenants', '/metrics', '/reports',
    '/room-mapping', '/sports-dashboard', '/sports-booking', '/hall-booking', '/fines',
    '/placements', '/analytics', '/leaves', '/room-requests', '/audit-logs', '/alumni',
  ],
  super_admin: [
    ...COMMON_PATHS,
    '/rooms', '/gate-passes', '/attendance', '/meals', '/gate-scans',
    '/colleges', '/tenants', '/metrics', '/reports', '/room-mapping',
    '/visitors', '/complaints', '/leaves', '/sports-dashboard', '/sports-booking', '/hall-booking',
    '/placements', '/analytics', '/fines', '/events', '/room-requests', '/audit-logs', '/alumni',
  ],
  chef: [
      ...COMMON_PATHS,
      '/meals',
      '/attendance',
      '/complaints',
      '/sports-booking',
  ],
  head_chef: [
      ...COMMON_PATHS,
      '/meals',
      '/attendance',
      '/complaints',
      '/analytics',
      '/sports-booking',
  ],
  gate_security: [
      ...COMMON_PATHS,
      '/gate-passes',
      '/gate-scans',
      '/visitors',
      '/fines',
      '/sports-booking',
  ],
  security_head: [
      ...COMMON_PATHS,
      '/gate-passes',
      '/gate-scans',
      '/reports',
      '/metrics',
      '/visitors',
      '/fines',
      '/analytics',
      '/sports-booking',
  ],
  hr: [
    ...COMMON_PATHS,
    '/rooms', '/complaints', '/reports', '/tenants', '/attendance', '/placements', '/room-requests',
    '/fines', '/sports-booking',
  ],
  pd: [
    ...COMMON_PATHS,
    '/events', '/sports-booking', '/sports-dashboard', '/hall-booking',
    '/placements', '/fines',
  ],
  pt: [
    ...COMMON_PATHS,
    '/events',
    '/sports-booking',
    '/sports-dashboard',
    '/fines',
  ],
  principal: [
    ...COMMON_PATHS,
    '/reports', '/metrics', '/analytics', '/hall-booking', '/fines', '/placements',
    '/events', '/sports-booking',
  ],
  director: [
    ...COMMON_PATHS,
    '/reports',
    '/hall-booking',
    '/sports-booking',
  ],
  hod: [
    ...COMMON_PATHS,
    '/events',
    '/hall-booking',
    '/fines',
    '/sports-booking',
  ],
  incharge: [
    ...COMMON_PATHS,
    '/rooms',
    '/sports-booking',
    '/gate-passes',
  ],
}

const DAY_SCHOLAR_RESTRICTED_PATHS = [
  '/rooms',
  '/meals',
  '/gate-passes',
  '/leaves',
  '/visitors',
  '/room-mapping',
  '/room-requests'
]

export function getRoleHome(role?: string | null) {
  if (!role) return '/dashboard'
  const normalized = role.toLowerCase() as Role
  return ROLE_HOME[normalized] ?? '/dashboard'
}

export function canAccessPath(
  role: string | null | undefined, 
  path: string, 
  studentType?: string | null,
  isStudentHR?: boolean | null,
) {
  // If no role but path is common, allow (prevents flash on refresh)
  const isCommon = COMMON_PATHS.some(cp => path === cp || path.startsWith(`${cp}/`))
  if (isCommon) return true

  if (!role) return false
  const normalized = role.toLowerCase() as Role

  if (normalized === 'student' && isStudentHR) {
    if (path === '/room-mapping' || path.startsWith('/room-mapping/')) {
      return true
    }
  }

  // Student Type Restrictions (Day Scholar vs Hosteller)
  if (normalized === 'student' && studentType === 'day_scholar') {
    if (DAY_SCHOLAR_RESTRICTED_PATHS.some(p => path === p || path.startsWith(`${p}/`))) {
      return false
    }
  }

  const allowed = ROLE_ALLOWED_PATHS[normalized]
  if (!allowed) return false

  if (path === '/' || path === '') return true

  return allowed.some((route) => path === route || path.startsWith(`${route}/`))
}

File: backend_django/core/rbac.py
"""Central RBAC engine.

Permission data is read from the database (apps.rbac tables) and cached per
user.  The static PERMISSION_MATRIX below acts as a **safe fallback** in
case the DB tables have not been seeded yet (e.g. during initial migrations).

Public API
----------
    has_module_permission(user, module, capability)  -> bool
    get_user_capabilities(user, module)              -> Set[str]
    get_path_grants_for_user(user)                   -> list[str]
"""

from __future__ import annotations

from typing import Dict, List, Set

from django.core.cache import cache

from core import cache_keys as ck
from core.constants import (
    ROLE_ADMIN,
    ROLE_CHEF,
    ROLE_DIRECTOR,
    ROLE_GATE_SECURITY,
    ROLE_HEAD_CHEF,
    ROLE_HEAD_WARDEN,
    ROLE_HOD,
    ROLE_INCHARGE,
    ROLE_PD,
    ROLE_PRINCIPAL,
    ROLE_PT,
    ROLE_SECURITY_HEAD,
    ROLE_STUDENT,
    ROLE_SUPER_ADMIN,
    ROLE_WARDEN,
    TOP_LEVEL_ROLES as TOP_LEVEL_ROLE_LIST,
)

MODULE_HOSTEL = 'hostel'
MODULE_SPORTS = 'sports'
MODULE_HALL = 'hall'
MODULE_FEES = 'fees'
MODULE_GATEPASS = 'gatepass'
MODULE_NOTICES = 'notices'
MODULE_MEALS = 'meals'
MODULE_SECURITY = 'security'
MODULE_REPORTS = 'reports'
MODULE_COMPLAINTS = 'complaints'
MODULE_NOTIFICATIONS = 'notifications'

ROLE_SECURITY = 'security'


ROLE_ALIASES = {
    ROLE_GATE_SECURITY: ROLE_SECURITY,
    ROLE_SECURITY_HEAD: ROLE_SECURITY,
}

TOP_LEVEL_ROLES = set(TOP_LEVEL_ROLE_LIST)
WARDEN_LEVEL_ROLES = TOP_LEVEL_ROLES | {ROLE_WARDEN}
GLOBAL_OWNER_ROLES = {ROLE_SUPER_ADMIN}
COLLEGE_OWNER_ROLES = {ROLE_ADMIN}

ROLE_GOVERNANCE_PROFILE: Dict[str, Dict[str, str]] = {
    ROLE_SUPER_ADMIN: {
        'scope': 'global',
        'label': 'Super Admin Owner',
        'description': 'Full product owner across all tenants/colleges.',
    },
    ROLE_ADMIN: {
        'scope': 'college',
        'label': 'College Admin Owner',
        'description': 'College-level owner with full control inside assigned college.',
    },
    ROLE_HEAD_WARDEN: {
        'scope': 'college_hostel',
        'label': 'Head Warden',
        'description': 'Head of warden operations and hostel governance within college.',
    },
    ROLE_WARDEN: {
        'scope': 'building_or_floor',
        'label': 'Warden',
        'description': 'Operational control for assigned blocks/floors under head-warden governance.',
    },
    ROLE_HEAD_CHEF: {
        'scope': 'college_kitchen',
        'label': 'Head Chef',
        'description': 'Leads kitchen/mess workflows and supervises chef operations.',
    },
    ROLE_CHEF: {
        'scope': 'kitchen_operations',
        'label': 'Chef',
        'description': 'Daily menu and meal operations under head-chef direction.',
    },
    ROLE_SECURITY_HEAD: {
        'scope': 'college_security',
        'label': 'Head Security',
        'description': 'Leads security operations and supervises gate-security workflows.',
    },
    ROLE_GATE_SECURITY: {
        'scope': 'security_execution',
        'label': 'Gate Security',
        'description': 'Executes scan/verification operations assigned by head security.',
    },
    ROLE_INCHARGE: {
        'scope': 'assigned_partial',
        'label': 'Incharge',
        'description': 'Limited assigned-scope support role without full management authority.',
    },
}


def _extract_role(role_or_user) -> str | None:
    """Accept either a role string or a user-like object."""
    if isinstance(role_or_user, str):
        return role_or_user
    return getattr(role_or_user, 'role', None)


def is_top_level_management(role_or_user) -> bool:
    """Backward-compatible top-level access check used by older modules."""
    role = _extract_role(role_or_user)
    return role in TOP_LEVEL_ROLES


def is_warden(role_or_user) -> bool:
    """Backward-compatible warden-or-higher access check used by older modules."""
    role = _extract_role(role_or_user)
    return role in WARDEN_LEVEL_ROLES


PERMISSION_MATRIX: Dict[str, Dict[str, str]] = {
    ROLE_SUPER_ADMIN: {
        MODULE_HOSTEL: 'full',
        MODULE_SPORTS: 'full',
        MODULE_HALL: 'full',
        MODULE_FEES: 'full',
        MODULE_GATEPASS: 'full',
        MODULE_NOTICES: 'full',
        MODULE_MEALS: 'full',
        MODULE_SECURITY: 'full',
        MODULE_REPORTS: 'full',
        MODULE_COMPLAINTS: 'full',
        MODULE_NOTIFICATIONS: 'full',
    },
    ROLE_ADMIN: {
        MODULE_HOSTEL: 'full',
        MODULE_SPORTS: 'full',
        MODULE_HALL: 'full',
        MODULE_FEES: 'full',
        MODULE_GATEPASS: 'full',
        MODULE_NOTICES: 'full',
        MODULE_MEALS: 'full',
        MODULE_SECURITY: 'full',
        MODULE_REPORTS: 'full',
        MODULE_COMPLAINTS: 'full',
        MODULE_NOTIFICATIONS: 'full',
    },
    ROLE_PRINCIPAL: {
        MODULE_HOSTEL: 'view',
        MODULE_SPORTS: 'view',
        MODULE_HALL: 'manage',
        MODULE_FEES: 'reports',
        MODULE_GATEPASS: 'view',
        MODULE_NOTICES: 'broadcast',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'view',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'view',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_DIRECTOR: {
        MODULE_HOSTEL: 'view',
        MODULE_SPORTS: 'view',
        MODULE_HALL: 'manage',
        MODULE_FEES: 'view',
        MODULE_GATEPASS: 'view',
        MODULE_NOTICES: 'broadcast',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'view',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'view',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_HOD: {
        MODULE_HOSTEL: 'none',
        MODULE_SPORTS: 'apply_for_class_branch',
        MODULE_HALL: 'request',
        MODULE_FEES: 'view',
        MODULE_GATEPASS: 'none',
        MODULE_NOTICES: 'department_notices',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'none',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'view',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_WARDEN: {
        MODULE_HOSTEL: 'manage',
        MODULE_SPORTS: 'none',
        MODULE_HALL: 'none',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'approve',
        MODULE_NOTICES: 'hostel_notices',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'view',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'manage',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_INCHARGE: {
        MODULE_HOSTEL: 'view',
        MODULE_SPORTS: 'view',
        MODULE_HALL: 'none',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'view',
        MODULE_NOTICES: 'create',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'none',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'view',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_PD: {
        MODULE_HOSTEL: 'none',
        MODULE_SPORTS: 'full_control',
        MODULE_HALL: 'manage',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'none',
        MODULE_NOTICES: 'sports_notices',
        MODULE_MEALS: 'none',
        MODULE_SECURITY: 'none',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'none',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_PT: {
        MODULE_HOSTEL: 'none',
        MODULE_SPORTS: 'assist_verify',
        MODULE_HALL: 'none',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'none',
        MODULE_NOTICES: 'none',
        MODULE_MEALS: 'none',
        MODULE_SECURITY: 'none',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'none',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_SECURITY: {
        MODULE_HOSTEL: 'none',
        MODULE_SPORTS: 'none',
        MODULE_HALL: 'none',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'verify_entry_exit',
        MODULE_NOTICES: 'none',
        MODULE_MEALS: 'none',
        MODULE_SECURITY: 'manage',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'view',
        MODULE_NOTIFICATIONS: 'view',
    },
    ROLE_STUDENT: {
        MODULE_HOSTEL: 'limited',
        MODULE_SPORTS: 'participate',
        MODULE_HALL: 'none',
        MODULE_FEES: 'view',
        MODULE_GATEPASS: 'request',
        MODULE_NOTICES: 'view',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'none',
        MODULE_REPORTS: 'none',
        MODULE_COMPLAINTS: 'create',
        MODULE_NOTIFICATIONS: 'view',
    },
    # HR role — scoped hostel/gatepass view, attendance management
    'hr': {
        MODULE_HOSTEL: 'manage',
        MODULE_SPORTS: 'none',
        MODULE_HALL: 'none',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'approve',
        MODULE_NOTICES: 'hostel_notices',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'none',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'view',
        MODULE_NOTIFICATIONS: 'view',
    },
    # head_warden — same as warden but with metrics access (handled via ROLE_EXTRA_PATHS)
    'head_warden': {
        MODULE_HOSTEL: 'manage',
        MODULE_SPORTS: 'none',
        MODULE_HALL: 'none',
        MODULE_FEES: 'none',
        MODULE_GATEPASS: 'manage',
        MODULE_NOTICES: 'manage',
        MODULE_MEALS: 'view',
        MODULE_SECURITY: 'view',
        MODULE_REPORTS: 'view',
        MODULE_COMPLAINTS: 'manage',
        MODULE_NOTIFICATIONS: 'view',
    },
}


LEVEL_CAPABILITIES: Dict[str, Set[str]] = {
    'none':                   set(),
    'view':                   {'view'},
    'limited':                {'view', 'limited'},
    'reports':                {'view', 'reports'},
    'request':                {'view', 'request'},
    'participate':            {'view', 'participate'},
    'apply_for_class_branch': {'view', 'participate', 'apply'},
    'assist_verify':          {'view', 'verify'},
    'request_only':           {'request'},
    'partial':                {'view', 'partial'},
    'create':                 {'view', 'create'},
    'approve':                {'view', 'approve'},
    'verify_entry_exit':      {'view', 'verify'},
    'broadcast':              {'view', 'create', 'broadcast'},
    'department_notices':     {'view', 'create', 'department_scope'},
    'hostel_notices':         {'view', 'create', 'hostel_scope'},
    'sports_notices':         {'view', 'create', 'sports_scope'},
    'manage':     {'view', 'manage', 'create', 'update', 'delete', 'approve', 'verify', 'request'},
    'full_control': {'view', 'manage', 'create', 'update', 'delete', 'approve', 'verify', 'request'},
    'full': {'view', 'manage', 'create', 'update', 'delete', 'approve', 'verify',
             'request', 'reports', 'broadcast'},
}

# ---------------------------------------------------------------------------
# Path grants: module + permission-level → frontend routes
# ---------------------------------------------------------------------------

COMMON_PATHS: List[str] = [
    '/dashboard', '/profile', '/notifications', '/messages', '/notices',
    '/events', '/digital-id', '/fines', '/disciplinary', '/unauthorized',
]

MODULE_PATH_GRANTS: Dict[str, Dict[str, List[str]]] = {
    MODULE_HOSTEL: {
        'none':     [],
        # students use hostel self-service pages; backend queryset scoping keeps data private
        'limited':  ['/rooms', '/attendance', '/leaves', '/visitors'],
        # principals/directors read hostel overview via /reports — no /rooms
        'view':     [],
        'partial':  ['/rooms'],
        'manage':   ['/rooms', '/tenants', '/room-mapping', '/attendance', '/meals', '/complaints'],
        'full':     ['/rooms', '/tenants', '/room-mapping', '/attendance', '/meals', '/complaints'],
    },
    MODULE_SPORTS: {
        'none':                   [],
        'view':                   ['/events', '/sports-booking'],
        'participate':            ['/events', '/sports-booking'],
        'apply_for_class_branch': ['/events', '/sports-booking'],
        'partial':                ['/events', '/sports-booking', '/sports-dashboard'],
        'assist_verify':          ['/events', '/sports-booking', '/sports-dashboard'],
        'full_control':           ['/events', '/sports-booking', '/sports-dashboard'],
        'manage':                 ['/events', '/sports-booking', '/sports-dashboard'],
        'full':                   ['/events', '/sports-booking', '/sports-dashboard'],
    },
    MODULE_HALL: {
        'none':    [],
        'request': ['/hall-booking'],
        'manage':  ['/hall-booking'],
        'full':    ['/hall-booking'],
    },
    MODULE_FEES: {
        'none':    [],
        'limited': ['/fines'],
        'view':    ['/fines'],
        'reports': ['/fines'],
        'manage':  ['/fines'],
        'full':    ['/fines'],
    },
    MODULE_GATEPASS: {
        'none':               [],
        'view':               ['/gate-passes'],
        'request':            ['/gate-passes'],
        'partial':            ['/gate-passes'],
        # approve → wardens/HR can also manage leaves and see visitors
        'approve':            ['/gate-passes', '/gate-scans', '/leaves', '/visitors'],
        'verify_entry_exit':  ['/gate-passes', '/gate-scans', '/visitors'],
        'manage':             ['/gate-passes', '/gate-scans', '/visitors', '/leaves'],
        'full':               ['/gate-passes', '/gate-scans', '/visitors', '/leaves'],
    },
    MODULE_NOTICES: {
        'none':               [],
        'view':               ['/notices'],
        'create':             ['/notices'],
        'broadcast':          ['/notices'],
        'department_notices': ['/notices'],
        'hostel_notices':     ['/notices'],
        'sports_notices':     ['/notices'],
        'manage':             ['/notices'],
        'full':               ['/notices'],
    },
    MODULE_MEALS: {
        'none':    [],
        'view':    ['/meals'],
        'create':  ['/meals'],
        'manage':  ['/meals'],
        'full':    ['/meals'],
    },
    MODULE_SECURITY: {
        'none':               [],
        'view':               ['/gate-scans', '/visitors'],
        'verify_entry_exit':  ['/gate-scans', '/visitors'],
        'manage':             ['/gate-scans', '/visitors'],
        'full':               ['/gate-scans', '/visitors'],
    },
    MODULE_REPORTS: {
        'none':     [],
        'view':     ['/reports'],
        'reports':  ['/reports'],
        'manage':   ['/reports'],
        'full':     ['/reports'],
    },
    MODULE_COMPLAINTS: {
        'none':    [],
        'view':    ['/complaints'],
        'create':  ['/complaints'],
        'manage':  ['/complaints'],
        'full':    ['/complaints'],
    },
    MODULE_NOTIFICATIONS: {
        'none':    [],
        'view':    ['/notifications'],
        'create':  ['/notifications'],
        'manage':  ['/notifications'],
        'full':    ['/notifications'],
    },
}

# Paths not derivable from module permissions alone (role-specific extras)
ROLE_EXTRA_PATHS: Dict[str, List[str]] = {
    'super_admin':   ['/colleges', '/metrics', '/reports', '/users'],
    'admin':         ['/colleges', '/metrics', '/reports', '/users'],
    'head_warden':   ['/metrics', '/reports', '/meals'],
    'warden':        ['/metrics', '/reports', '/meals'],
    'principal':     ['/reports'],
    'director':      ['/reports'],
    'security_head': ['/reports', '/metrics'],
    'chef':          ['/meals', '/attendance', '/complaints'],
    'head_chef':     ['/meals', '/attendance', '/complaints'],
    'hr':            ['/reports', '/attendance'],
    'gate_security': ['/visitors'],
    'student':       ['/resume', '/placements', '/hall-booking'],
}

DAY_SCHOLAR_RESTRICTED_PATHS: List[str] = [
    '/rooms', '/meals', '/gate-passes', '/leaves', '/visitors', '/room-mapping',
]


def normalize_role(role: str | None) -> str:
    if not role:
        return ''
    role = role.lower()
    return ROLE_ALIASES.get(role, role)


def get_role_governance_profile(role_or_user) -> Dict[str, str]:
    """Return human-readable governance metadata for role clarity in clients."""
    role = normalize_role(_extract_role(role_or_user))
    return ROLE_GOVERNANCE_PROFILE.get(
        role,
        {
            'scope': 'standard',
            'label': role or 'unknown',
            'description': 'Standard role with module-level permissions from RBAC.',
        },
    )


def _permissions_cache_key(user_id: int) -> str:
    return ck.permissions_user(user_id)


def get_user_module_levels(user) -> Dict[str, str]:
    """Return {module_slug: permission_level_slug} for the user's role.

    Data source priority:
      1. In-process cache (Redis / local-mem)
      2. Database RolePermission table  (seeded via ``manage.py seed_rbac``)
      3. Static PERMISSION_MATRIX fallback (un-seeded / migration time)
    """
    if not getattr(user, 'is_authenticated', False):
        return {}

    cache_key = _permissions_cache_key(user.id)
    cached = cache.get(cache_key)
    if cached is not None:
        return cached

    role = normalize_role(getattr(user, 'role', None))

    # ── Try DB first ─────────────────────────────────────────────────────────
    try:
        from apps.rbac.models import RolePermission
        rows = (
            RolePermission.objects
            .filter(role__slug=role)
            .select_related('module', 'permission')
            .values('module__slug', 'permission__slug')
        )
        if rows.exists():
            role_matrix = {
                row['module__slug']: row['permission__slug']
                for row in rows
            }
            
            # --- DYNAMIC HR PERMISSIONS INJECTION ---
            if getattr(user, 'is_hr', False):
                try:
                    hr_perm = getattr(user, 'hr_permissions', None)
                    if hr_perm:
                        if getattr(hr_perm, 'can_view_attendance', False):
                            role_matrix[MODULE_HOSTEL] = 'view'
                        if getattr(hr_perm, 'can_view_complaints', False):
                            role_matrix[MODULE_COMPLAINTS] = 'view'
                        elif getattr(hr_perm, 'can_raise_complaints', True):
                            if role_matrix.get(MODULE_COMPLAINTS, 'none') == 'none':
                                role_matrix[MODULE_COMPLAINTS] = 'create'
                        if getattr(hr_perm, 'can_assist_reports', False):
                            role_matrix[MODULE_REPORTS] = 'view'
                        if getattr(hr_perm, 'can_manage_notices', False):
                            role_matrix[MODULE_NOTICES] = 'hostel_notices'
                        
                        # ENFORCEMENT: HR can NEVER override gate pass, fine, wardens
                        role_matrix[MODULE_GATEPASS] = 'request'  # Student tier, no approve
                        role_matrix[MODULE_FEES] = 'view'         # Fines view only, no issue
                except Exception:
                    pass
            # --- END DYNAMIC HR ---
            
            cache.set(cache_key, role_matrix, timeout=900)
            return role_matrix

    except Exception:
        pass  # DB not ready (makemigrations context) — fall through

    # ── Static fallback ──────────────────────────────────────────────────────
    role_matrix = PERMISSION_MATRIX.get(role, {})
    if role_matrix:
        import logging
        logging.getLogger(__name__).debug(
            "RBAC: using static PERMISSION_MATRIX fallback for role '%s' (user_id=%s). "
            "Run 'manage.py seed_rbac' to populate the DB.",
            role, getattr(user, 'id', '?'),
        )
    cache.set(cache_key, role_matrix, timeout=900)
    return role_matrix


def get_user_capabilities(user, module: str) -> Set[str]:
    levels = get_user_module_levels(user)
    level = levels.get(module, 'none')
    return LEVEL_CAPABILITIES.get(level, set())


def has_module_permission(user, module: str, capability: str) -> bool:
    if not getattr(user, 'is_authenticated', False):
        return False
    if getattr(user, 'is_superuser', False):
        return True

    capabilities = get_user_capabilities(user, module)
    if capability == 'view' and 'limited' in capabilities:
        return True

    return capability in capabilities or 'full' in capabilities or 'manage' in capabilities


# ---------------------------------------------------------------------------
# Path grant computation (used by /auth/my-permissions/ endpoint)
# ---------------------------------------------------------------------------

def get_path_grants_for_user(
    user,
    module_levels: Dict[str, str] | None = None,
) -> List[str]:
    """Return the sorted, deduplicated list of frontend route paths the user
    is allowed to visit.  Keys match the ``href`` values in Sidebar.tsx.
    """
    if module_levels is None:
        module_levels = get_user_module_levels(user)

    role = normalize_role(getattr(user, 'role', None))
    seen: set[str] = set()
    paths: List[str] = []

    def _add(p: str) -> None:
        if p not in seen:
            seen.add(p)
            paths.append(p)

    # 1. Common paths — always included
    for p in COMMON_PATHS:
        _add(p)

    # 2. Module-driven paths
    for module, level in module_levels.items():
        for p in MODULE_PATH_GRANTS.get(module, {}).get(level, []):
            _add(p)

    # 3. Role-specific extras not covered by modules
    for p in ROLE_EXTRA_PATHS.get(role, []):
        _add(p)

    if role == ROLE_STUDENT and getattr(user, 'is_student_hr', False):
        _add('/room-mapping')

    # 4. Day-scholar restrictions
    student_type = getattr(user, 'student_type', '')
    if role == ROLE_STUDENT and student_type == 'day_scholar':
        restricted_paths = set(DAY_SCHOLAR_RESTRICTED_PATHS)
        if getattr(user, 'is_student_hr', False):
            restricted_paths.discard('/room-mapping')
        paths = [p for p in paths if p not in restricted_paths]

    return sorted(set(paths))


# ---------------------------------------------------------------------------
# Cache management
# ---------------------------------------------------------------------------

def clear_user_permission_cache(user_id: int) -> None:
    cache.delete(_permissions_cache_key(user_id))
